File: src/main/scala/me/ktar5/infoboard/InfoBoard.scala
package me.ktar5.infoboard

import java.io.IOException

import me.clip.placeholderapi.PlaceholderAPI
import me.ktar5.infoboard.logger.Logger
import net.milkbowl.vault.chat.Chat
import net.milkbowl.vault.economy.Economy
import net.milkbowl.vault.permission.Permission
import org.bukkit.Bukkit
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.{EventHandler, EventPriority}
import org.bukkit.plugin.java.JavaPlugin
import org.mcstats.Metrics

/*
 * Copyright (C) 2013-Current Carter Gale (Ktar5) <buildfresh@gmail.com>
 * 
 * This file is part of infoboard.
 * 
 * infoboard can not be copied and/or distributed without the express
 * permission of the aforementioned owner.
 */

class InfoBoard extends JavaPlugin{


    override def onEnable(): Unit = {
        InfoBoard._instance = this

        if(!InfoBoard.vault.init()){
            InfoBoard.logger.warn("Hey! Vault is not enabled, here be dragons")
        }

        try {
            val metrics = new Metrics(this)
            metrics.start()
        } catch {
            case ioe: IOException => InfoBoard.logger.fatal("Hey! Metrics were unable to start... this is odd")
        }

        if (Bukkit.getPluginManager.isPluginEnabled("PlaceholderAPI")) {
            //todo
        } else {
            //todo
            InfoBoard.logger.fatal("Could not find PlaceholderAPI!! Plugin can not work without it!")
            throw new RuntimeException("Could not find PlaceholderAPI!! Plugin can not work without it!")
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    def onJoin(event: PlayerJoinEvent) {

        var withoutPlaceholdersSet = "%player_name% &ajoined the server! He/she is rank &f%vault_rank%";

        var withPlaceholdersSet = PlaceholderAPI.setPlaceholders(event.getPlayer(), withoutPlaceholdersSet);

        event.setJoinMessage(withPlaceholdersSet);
    }

}

object InfoBoard{

    val vault = new Vault
    val logger: Logger = new Logger

    private var _instance: InfoBoard = null
    def instance = _instance

    class Vault {
        private var _econ: Economy = null
        private var _perms: Permission = null
        private var _chat: Chat = null

        def econ = _econ
        def perms = _perms
        def chat = _chat

        def init(): Boolean ={
            if (instance.getServer.getPluginManager.getPlugin("Vault") == null) {
                return false
            }
            _econ = instance.getServer.getServicesManager.getRegistration[Economy](Class[Economy]).getProvider
            _chat = instance.getServer.getServicesManager.getRegistration[Chat](Class[Chat]).getProvider
            _perms = instance.getServer.getServicesManager.getRegistration[Permission](Class[Permission]).getProvider
            true
        }

    }

}

File: src/main/java/com/ktar5/infoboard/InfoBoard.java
package com.ktar5.infoboard;

import com.ktar5.infoboard.API.Vault;
import com.ktar5.infoboard.utils.FileManager;
import com.ktar5.infoboard.utils.Metrics;
import net.milkbowl.vault.economy.Economy;
import net.milkbowl.vault.permission.Permission;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scoreboard.DisplaySlot;

import java.io.IOException;
import java.util.ArrayList;

public class InfoBoard extends JavaPlugin {

    public static Plugin instance;
    public static Economy economy;
    public static Permission permission;
    public static boolean economyB;
    public static boolean permissionB;
    public static ArrayList<String> hidefrom = new ArrayList<String>();
    private static Timers timers;
    private static FileManager fileManager;
    public boolean update = false;
    public String name = "InfoBoard";
    public String ib = "" + ChatColor.RED + ChatColor.BOLD + "➳"
            + ChatColor.GRAY;

    /**
     * @return the fileManager
     */
    public static FileManager getFileManager() {
        return InfoBoard.fileManager;
    }

    /**
     * @return the timer
     */
    public static Timers getTimers() {
        return InfoBoard.timers;
    }

    /**
     * Load Metrics
     */
    private void loadMetrics() {
        try {
            Metrics metrics = new Metrics(this);
            metrics.start();
            System.out.println("Metrics was started!");
        } catch (IOException e) {
            System.out.println("Metrics was unable to start...");
        }
    }

    @Override
    public void onDisable() {
        Bukkit.getScheduler().cancelTasks(this);
        for (Player player : Bukkit.getOnlinePlayers())
            if (player.getScoreboard().getObjective(DisplaySlot.SIDEBAR) != null)
                if (player.getScoreboard().getObjective(DisplaySlot.SIDEBAR)
                        .getName().equalsIgnoreCase("InfoBoard"))
                    player.getScoreboard().clearSlot(DisplaySlot.SIDEBAR);

        instance = null;
    }

    @Override
    public void onEnable() {
        if (!Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI")) {
            throw new RuntimeException("Could not find PlaceholderAPI!! Plugin can not work without it!");
        }
        if (!Bukkit.getPluginManager().isPluginEnabled("WorldGuard")) {
            throw new RuntimeException("Could not find WorldGuard!! Plugin can not work without it!");
        }
        if (!Bukkit.getPluginManager().isPluginEnabled("Vault")) {
            throw new RuntimeException("Could not find Vault!! Plugin can not work without it!");
        }
        InfoBoard.instance = this;
        InfoBoard.fileManager = new FileManager();

        InfoBoard.timers = new Timers();
        getTimers().start();

        Vault.load();
        loadMetrics();

        PluginManager pm = getServer().getPluginManager();
        pm.registerEvents(new PlayerListener(this), this);
        getCommand("InfoBoard").setExecutor(new Commands(this));

    }

}
